atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java

```diff
@@ -1024,14 +1074,21 @@ private CfDef getCfForTable(TableReference tableRef, byte[] rawMetadata) {
                                     if (shouldTraceQuery(tableRef)) {
                                         ByteBuffer recv_trace = client.trace_next_query();
                                         Stopwatch stopwatch = Stopwatch.createStarted();
-                                        firstPage = client.get_range_slices(colFam, pred, keyRange, consistency);
-                                        long duration = stopwatch.elapsed(TimeUnit.MILLISECONDS);
-                                        if (duration > getMinimumDurationToTraceMillis()) {
-                                            log.error("Traced a call to " + tableRef.getQualifiedName() + " that took " + duration + " ms."
-                                                    + " It will appear in system_traces with UUID=" + CassandraKeyValueServices.convertCassandraByteBufferUUIDtoString(recv_trace));
+                                        try {
+                                            firstPage = client.get_range_slices(colFam, pred, keyRange, consistency);
+                                        } catch (Exception e) {
+                                            logFailedCall(tableRef);
+                                            logTraceResults(stopwatch.elapsed(TimeUnit.MILLISECONDS), tableRef, recv_trace, true);
+                                            throw e;
                                         }
+                                        logTraceResults(stopwatch.elapsed(TimeUnit.MILLISECONDS), tableRef, recv_trace, false);
                                     } else {
-                                        firstPage = client.get_range_slices(colFam, pred, keyRange, consistency);
```

Comment:
I have a work in progress of this in my local stash from long ago if you're interested, but the gist of it is (needs updating for your tracing and switch to table refs:

```
    private interface Action<V> {
        V run() throws TException;
    }

    private void batchMutateInternal(final Client client,
                                    final Set<String> tableNames,
                                    final Map<ByteBuffer, Map<String, List<Mutation>>> map,
                                    final ConsistencyLevel consistency) throws TException {

        run(client, tableNames, new Action<Void>() {
            @Override
            public Void run() throws TException {
                client.batch_mutate(map, consistency);
                return null;
            }
        });
    }

    private <V> V run(Client client, Set<String> tableNames, Action<V> action) throws TException {
        if (shouldTraceQuery(tableNames)) {
            return trace(action, client, tableNames);
        } else {
            return action.run();
        }
    }

    private <V> V run(Client client, String tableName, Action<V> action) throws TException {
        if (shouldTraceQuery(tableName)) {
            return trace(action, client, ImmutableSet.of(tableName));
        } else {
            return action.run();
        }
    }

    private <V> V trace(Action<V> action, Client client, Set<String> tableNames) throws TException {
        ByteBuffer traceId = client.trace_next_query();
        Stopwatch stopwatch = Stopwatch.createStarted();
        try {
            return action.run();
        } finally {
            long duration = stopwatch.elapsed(TimeUnit.MILLISECONDS);
            if (duration > getMinimumDurationToTraceMillis()) {
                log.error("Traced a call to " + tableNames + " that took " + duration + " ms."
                        + " It will appear in system_traces with UUID="
                        + CassandraKeyValueServices.convertCassandraByteBufferUUIDtoString(traceId));
            }
        }
    }
```